### ISSUE TITLE
[Epic] Improve precision of source/sink status reporting

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#20036]
### Feature request

We have three reports of the `mz_{source|sink}_status_history` tables presenting an imprecise accounting of a source/sink's status that's lead to user confusion:

- https://github.com/MaterializeInc/materialize/issues/18795
- https://github.com/MaterializeInc/materialize/issues/16580
- https://github.com/MaterializeInc/materialize/issues/16582

Fixing these issues has been challenging because of the current architecture. Each source/sink is responsible for reporting its own status from `clusterd`. That means that if `clusterd` goes down (because all replicas of a cluster have been dropped, or Kubernetes is having issues), no error is reported in the table.

We have a design doc that will address all of these issues by changing the architecture to instead have `environmentd` write status updates to the `mz_{source|sink}_status_history` tables. This will allow `environmentd` to update the status to "unknown" if it has trouble communicating with a cluster:

* #19792

cc @dseisun-materialize — just tracking/bundling the work that @rkrishn7 is already doing in #19792.

**Comment by guswynn** [comment]
@bkirwi @benesch 
https://github.com/MaterializeInc/materialize/pull/23222 originally partially implemented https://github.com/MaterializeInc/materialize/issues/16580, but I removed it because it ultimately revealed information about our k8s setup, which to me seemed effectively not interpretable by a user. Do we actually want this? any pod error that isn't reported by `mz_cluster_replica_statuses` is a bug on our end, right?

**Comment by bkirwi** [comment]
It seems broadly useful to me for "the cluster the source/sink is running on is totally busted" to show up in the source/sink status at some point, even if it's mentioned in replica statuses, if only to send people hunting less. (It is ~the sort of source-wide health issue that that table can effectively represent.)

But I don't feel strongly about it at all!

**Comment by benesch** [comment]
Post #23222, how does the issue present? Say I have a source whose cluster is in CrashLoopBackoff—the source keeps panicking right after startup, and so the backoff period gets longer and longer. Will I see "starting", "running", "stalled", (backoff occurs), "starting", "running", "stalled", (backoff occurs), "starting", "running", "stalled"? Or do I see "starting", "running", (backoff occurs), "starting", "running", (backoff occurs)?

Agree with not wanting to leak too much k8s state into the `source_status_history` relation. But agree with Ben that we want to reflect something in the source status history table—and I think that "something" is somehow ensuring that the status is "stalled" when the source replica's pod is in CrashLoopBackoff.

**Comment by guswynn** [comment]
@benesch I am not sure; i don't think we test that because we in general we don't have a thing like `mz_panic` to force a storage cluster to go down (though we could do so with that new unification feature flag)

definitely worth looking into telling the user that the cluster crashed, but we should be thoughtful about how to present it; unfortunately the `RehydratingStorageClient` makes this a bit annoying to manage...

**Comment by benesch** [comment]
>  i don't think we test that because we in general we don't have a thing like mz_panic to force a storage cluster to go down (though we could do so with that new unification feature flag)

Should be able to do it with mzcompose pretty easily! Either use `docker exec` to kill the `clusterd` process every few seconds, or use `REMOTE` replicas and `c.down`.

> but we should be thoughtful about how to present it; unfortunately the `RehydratingStorageClient` makes this a bit annoying to manage...

Can't you just have the `RehydratingStorageClient` conjure up a "SourceStatus(Stalled)" message every time it observes the client disconnect? The stalled can be something generic like "internal error: lost connection to source cluster".

**Comment by guswynn** [comment]
> Can't you just have the RehydratingStorageClient conjure up a "SourceStatus(Stalled)" message every time it observes the client disconnect?

right now sources start paused, with this change they would also start with this stall, unless the rehydrating client knows more about if its waiting on the _first_ connection i think?

**Comment by benesch** [comment]
Right, exactly! You want the rehydrating client to produce the stall message on a _disconnection event_, and not in the initial disconnected state.
